### PR TITLE
M5472 Specify how reference attributes are mapped

### PR DESCRIPTION
**Pull request by Mark-de-Haan**
### Feature
it is now possible to use an entity with all its columns inside a magma script expression.
[Test file](https://github.com/molgenis/molgenis/files/1855516/m5472_test_emx.xlsx)

```js
/*
 * XREF & CATEGORICAL
 */

// data
refColumn -> {id: '1', label: 'label 1', age: '25'}  

// old situation
$('refColumn').value() -> '1'

// new situation
$('refColumn').value() -> '1'
$('refColumn.id').value() -> '1'
$('refColumn.label').value() -> 'label 1'
$('refColumn.age').value() -> '25'

/*
 * MREF & CATEGORICAL_MREF
 */

// data
refColumn -> [{id: '1', label: 'label 1', age: '25'}, {id: '2', label: 'label 2', age: '50'}]

// old situation  
$('refColumn').value() -> ['1', '2']

// new situation
$('refColumn').map(function (ref) {
  return ref.value() -> ['1', '2']
  // or 
  return ref.val.id -> ['1', '2']
  // or
  return ref.val.label -> ['label 1', 'label 2']
  // or
  return ref.val.age -> ['25', '50']
})
```
#### Checklist
- [x] Functionality works & meets specifications
- [x] Code reviewed
- [x] Code unit/integration/system tested
- [x] User documentation updated
- [x] (If you have changed REST API interface) view-swagger.ftl updated
- [ ] Test plan template updated
- [x] Clean commits
- [x] Added Feature/Fix to release notes
- [x] Integration tests run correctly

### CODE REVIEW

**Comment by Mark-de-Haan** [comment]
test this please

**Comment by Mark-de-Haan** [comment]
test this please

File: docs/user_documentation/ref-expressions.md [hunk 1]
```diff
@@ -44,18 +45,84 @@ $('myIntAttributeName').gt(3).and($('myIntAttributeName').lt(6)).value()
...
 | map      | Object     | $('data').map({0:1, 1:2}).value()        | Maps categories to eachother                                                               |
 | group    | Array      | $('age').group([18, 35, 50, 75]).value() | Produces ranges which are left inclusive, (-∞, 18), [18, 35), [35, 50), [50, 75), [75, +∞)|                                                                |
+
 ## Terminal operations
 | Operator | Parameters | Example             | Description      |
 |----------|------------|---------------------|------------------|
 | value    | -          | $('Height').value() | JavaScript value |
+
+# Special case: reference types
+If an attribute is a refereence type (MREF, XREF, CATEGORICAL, CATEGORICAL_MREF) you can use dot notation, to access the values of different columns in the row being referenced. E.g. `$('cookie.name').value()` gives you the value of the name column inside the table being referenced by the person column. See below for a more detailed example.
```
Comment (ConnorStroomberg) [inline docs/user_documentation/ref-expressions.md]: 'refereence' typo ?

Comment (ConnorStroomberg) [inline docs/user_documentation/ref-expressions.md]: 'by the person column'  typo ? there is no person column

File: molgenis-js/src/main/java/org/molgenis/js/magma/JsMagmaScriptEvaluator.java [hunk 2]
```diff
@@ -53,7 +53,7 @@ public Object eval(String expression, Entity entity)
 			stopwatch = Stopwatch.createStarted();
 		}
 
-		Map<String, Object> scriptEngineValueMap = toScriptEngineValueMap(entity);
+		Object scriptEngineValueMap = toScriptEngineValueMap(entity, 3);
```
Comment (ConnorStroomberg) [inline molgenis-js/src/main/java/org/molgenis/js/magma/JsMagmaScriptEvaluator.java]: magic number ??

File: molgenis-js/src/main/java/org/molgenis/js/magma/JsMagmaScriptEvaluator.java [hunk 3]
```diff
@@ -73,16 +73,39 @@ public Object eval(String expression, Entity entity)
 		return value;
 	}
 
-	private static Map<String, Object> toScriptEngineValueMap(Entity entity)
+	/**
+	 * Convert entity to a JavaScript object.
+	 * Adds "_idValue" as a special key to every level for quick access to the id value of an entity.
+	 *
+	 * @param entity The entity to be flattened, should start with non null entity
```
Comment (ConnorStroomberg) [inline molgenis-js/src/main/java/org/molgenis/js/magma/JsMagmaScriptEvaluator.java]: depth is missing in javadoc

File: molgenis-js/src/main/java/org/molgenis/js/magma/JsMagmaScriptEvaluator.java [hunk 4]
```diff
@@ -73,16 +73,39 @@ public Object eval(String expression, Entity entity)
...
+	 * @return flattened entity as key value map
```
Comment (ConnorStroomberg) [inline molgenis-js/src/main/java/org/molgenis/js/magma/JsMagmaScriptEvaluator.java]: its not 'flattened' any more right , it is copied until depth level is reached

File: molgenis-js/src/main/java/org/molgenis/js/magma/JsMagmaScriptEvaluator.java [hunk 5]
```diff
@@ -73,16 +73,39 @@ public Object eval(String expression, Entity entity)
...
+			{
+				return idValue;
+			}
+			else
+			{
+				Map<String, Object> map = Maps.newHashMap();
+				entity.getEntityType()
+					  .getAtomicAttributes()
+					  .forEach(attr -> map.put(attr.getName(), toScriptEngineValue(entity, attr, depth)));
+				map.put("_idValue", idValue);
```
Comment (ConnorStroomberg) [inline molgenis-js/src/main/java/org/molgenis/js/magma/JsMagmaScriptEvaluator.java]: should '_idValue' be added to the help pages as reserved keyword and or blacklisted as prop name ?

File: molgenis-js/src/main/java/org/molgenis/js/magma/JsMagmaScriptEvaluator.java [hunk 6]
```diff
@@ -97,16 +120,25 @@ private static Object toScriptEngineValue(Entity entity, Attribute attr)
...
+				{
+					jsArray = (ScriptObjectMirror) jsScriptEngine.eval("var arr = []; arr");
+					@SuppressWarnings("unchecked")
+					List<Object> mrefValues = jsArray.to(List.class);
+					entity.getEntities(attrName)
+						  .forEach(mrefEntity -> mrefValues.add(toScriptEngineValueMap(mrefEntity, depth - 1)));
+				}
+				catch (javax.script.ScriptException ex)
+				{
+					// ignore
```
Comment (ConnorStroomberg) [inline molgenis-js/src/main/java/org/molgenis/js/magma/JsMagmaScriptEvaluator.java]: is this a good idea ?, why not handle it , a the least add doc why this is ignored

File: molgenis-js/src/main/resources/js/script-evaluator.js [hunk 7]
```diff
@@ -20,11 +19,24 @@ function evalScript (script, entity) {
...
        * attribute
        *
+       * If the value is an object, return the idValue
+       *
        * @memberof $
        * @method value
        *
        */
       value: function () {
+        if (this.val !== null && typeof this.val === 'object') {
```
Comment (ConnorStroomberg) [inline molgenis-js/src/main/resources/js/script-evaluator.js]: really need a unit test of this ? or is there one already ? code cov does not pick find this one i would guess

File: molgenis-js/src/main/java/org/molgenis/js/magma/JsMagmaScriptEvaluator.java (hunk 5 above)
Comment (Mark-de-Haan) [reply molgenis-js/src/main/java/org/molgenis/js/magma/JsMagmaScriptEvaluator.java]: Yes

**Pull request merged**